services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java

```diff
@@ -0,0 +1,188 @@
+package com.futurewei.alcor.gatewaymanager.service;
+
+import com.futurewei.alcor.common.db.CacheException;
+import com.futurewei.alcor.common.executor.AsyncExecutor;
+import com.futurewei.alcor.gatewaymanager.config.ExceptionMsgConfig;
+import com.futurewei.alcor.gatewaymanager.dao.GWAttachmentRepository;
+import com.futurewei.alcor.gatewaymanager.dao.GatewayRepository;
+import com.futurewei.alcor.gatewaymanager.entity.GWAttachment;
+import com.futurewei.alcor.gatewaymanager.entity.ResourceType;
+import com.futurewei.alcor.gatewaymanager.entity.StatusEnum;
+import com.futurewei.alcor.web.entity.gateway.*;
+import com.futurewei.alcor.web.restclient.GatewayManagerRestClinet;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+import java.util.function.Supplier;
+
+@Slf4j
+@Service
+@ComponentScan("com.futurewei.alcor.web.restclient")
+public class GatewayService {
+
+    @Autowired
+    private GatewayRepository gatewayRepository;
+
+    @Autowired
+    private GWAttachmentRepository gwAttachmentRepository;
+
+    @Autowired
+    private GatewayManagerRestClinet restClinet;
+
+    private static final String GATEWAY_DESCRIPTION = "internal gateway";
+    private static final String GATEWAY_SUFFIX_NAME = " name";
+    private static final String ATTACHMENT_NAME_PREFIX = "attachment - ";
+
+
+    public GatewayService() {
+    }
+
+    public GatewayInfo createGatewayInfo(String projectId, VpcInfo vpcInfo) throws Exception {
+
+        //TODO check whether the project type is zeta by projectId
+
+        // create GatewayInfo entity in the DB
+        GatewayInfo gatewayInfo = new GatewayInfo();
+        GatewayEntity gatewayEntity = new GatewayEntity();
+
+        // construct the GatewayEntity and Attachment
+        createGatewayAndAttachment(projectId, vpcInfo, gatewayInfo, gatewayEntity);
+
+        AsyncExecutor executor = new AsyncExecutor();
+        executor.runAsync(restClinet::createDPMCacheGateway, projectId, gatewayInfo);
+        executor.runAsync(restClinet::createVPCInZetaGateway, new VpcInfoSub(vpcInfo.getVpcId(), vpcInfo.getVpcVni()));
+        executor.runAsync((Supplier<Void>) () -> {
+            // wait result of all async
+            List<Object> result = executor.joinAllAsync();
+            // whether to rollback by checking result's size is 2
+            try {
+                updateDPMCacheGateway(result, gatewayInfo, gatewayEntity, projectId);
+            } catch (Exception e) {
+                log.info("update GatewayEntity status to READY failed, detail message: {}", e.getMessage());
+                //TODO how to handle this exception? rollback all?
```

Comment:
If the execution goes to this far, that's mean DPM already has an gatewayinfo entry for zeta with status = PENDING. In that case, when port creation comes to DPM, DPM should wait for the status become READY for 30s. If the status still PENDING or FAILED, DPM will switch to OVS plugin. So, should we let the updateDPMCacheGateway try some times and if still failed rollback GM's DB and ask Zeta to remove the corresponding gateway resources.

Reply:
if update DPM's cache status to READY is fail,we need retry some times(3 times?),if still failed,we should rollback all the request resource, is this mean?

Reply:
Yes. that's what I think.